src/mug.gleam

```
@@ -171,7 +396,162 @@ fn gen_tcp_connect(
+///
+pub fn connect(options: ConnectionOptions) -> Result(Socket, Error) {
+  let host = charlist.from_string(options.host)
+  case options.tls_opts {
+    UseTls(TlsOptions(vm)) -> {
+      let opts =
+        erlang.rescue(fn() { get_tls_options(vm) })
+        |> result.replace_error(SystemCacertificatesGetError)
+      use opts <- result.try(opts)
+      ssl_connect(host, options.port, opts, options.timeout)
+      |> result.map(SslSocket)
+    }
+    _ -> {
+      let gen_options = [
+        // When data is received on the socket queue it in the TCP stack rather than
+        // sending it as an Erlang message to the socket owner's inbox.
+        #(Active, dynamic.from(False)),
+        // We want the data from the socket as bit arrays please, not lists.
+        #(Mode, dynamic.from(Binary)),
+      ]
+      gen_tcp_connect(host, options.port, gen_options, options.timeout)
+      |> result.map(TcpSocket)
+    }
+  }
+}
+
+@external(erlang, "mug_ffi", "ssl_upgrade")
+fn ssl_upgrade(
+  socket: TcpSocket,
+  options: List(SslOption),
+  timeout: Int,
+) -> Result(SslSocket, Error)
+
+/// Upgrade a plain TCP connection to TLS.
+///
+/// Accepts a socket and performs the client-side TLS handshake. This
+/// may not work on all TCP servers.
+///
+/// If the socket is already a TLS socket, this function does nothing.
+///
+/// Returns an error if the connection could not be established.
+///
+/// The socket is created in passive mode, meaning the the `receive` function is
+/// to be called to receive packets from the client. The
+/// `receive_next_packet_as_message` function can be used to switch the socket
+/// to active mode and receive the next packet as an Erlang message.
+///
+pub fn upgrade(
+  socket: Socket,
+  verification_method vm: TlsVerificationMethod,
+  timeout timeout: Int,
+) -> Result(Socket, Error) {
+  case socket {
+    TcpSocket(socket) -> {
+      let opts =
+        erlang.rescue(fn() { get_tls_options(vm) })
```

Never use rescue like this, it silently discards actual problems. Return errors from the Erlang instead.

---

Fixed!